.github/workflows/canister-tests.yml

```diff
@@ -195,26 +195,120 @@ jobs:
+        run: |
+          # Last time dependencies or toolchain were updated; used for ~/cargo and dependencies
+          commit_deps=$(git log -1 --pretty=format:%H -- Cargo.* src/*/Cargo.toml rust-toolchain.toml)
+          echo "::set-output name=commit_deps::$commit_deps"
+          echo "deps: $commit_deps"
+
+          # Last time the canister code was modified
+          commit_canister=$(git log -1 --pretty=format:%H -- src/[^frontend]*)
+          echo "::set-output name=commit_canister::$commit_canister"
+          echo "canister: $commit_canister"
+
+          # This commit
+          commit_now=$(git log -1 --pretty=format:%H)
+          echo "::set-output name=commit_now::$commit_now"
+          echo "now: $commit_now"
+
       - uses: ./.github/actions/bootstrap
 
       # Rustup only installs cargo on the first call, so we use a dedicated step to get a good idea of how
       # time is spent in each step separately
       - run: cargo check --help
 
-        # NOTE: the top-level crate "canister_tests" still gets recompiled (~4mn)
-        # due to the mtime changing on the checkout. Can be worked around by storing
-        # the original mtime in the cache and by hashing all of src/canister_tests
-        # (and then restoring the mtime).
+      - name: Create dummy assets
+        run: |
+          mkdir dist
+          touch dist/index.html
+          touch dist/index.js.gz
+          touch dist/index.css
+          touch dist/loader.webp
+          touch dist/favicon.ico
+          touch dist/ic-badge.svg
+
+      # Caching ~/.cargo
+
       - uses: actions/cache@v3
+        id: cache-dot-cargo
         with:
-          # note on paths: we include the whole ~/.cargo/. (as opposed to just .cargo/git, etc.) because cargo stores crate
-          # data as .crates(2?).{json,toml} without which the whole cache is invalidated. We may accumulate unnecessary
-          # stuff but the cache is anyway invalidated on every toolchain update.
-          # We also include all Cargo tomls and the lockfile
-          path: |
-            ~/.cargo
-            target
-          key: ${{ runner.os }}-cargo-${{ hashFiles('Cargo.lock', 'Cargo.toml', 'src/**/Cargo.toml', 'rust-toolchain.toml') }}-3
+          path: /tmp/dot-cargo
+          key: ${{ runner.os }}-${{ steps.git_info.outputs.commit_deps }}-dot-cargo
+
+      - run: git checkout ${{ steps.git_info.outputs.commit_deps }}
+
+      - name: Cargo cache (recovery)
+        if: steps.cache-dot-cargo.outputs.cache-hit == 'true'
+        run: |
+          rsync -av "/tmp/dot-cargo/" "$HOME/.cargo/"
+          find "$HOME/.cargo/" -exec touch -t "$(date +%Y%m%d%H%M.%S)" {} +
+
+      - name: Cargo cache (build)
+        run: cargo fetch --locked
+
+      - name: Cargo cache (save to cache)
+        if: steps.cache-dot-cargo.outputs.cache-hit != 'true'
+        run: rsync -av "$HOME/.cargo/" "/tmp/dot-cargo/"
+
+      # Caching external dependencies
+
+      - uses: actions/cache@v3
```

Comment:
This cache action is also dependent on `steps.git_info.outputs.commit_deps`, so in theory it could be merged with `cache-dot-cargo` and using multiple folders, right?

Reply:
True! However I've separated things as much as possible for legibility and also in case we make those steps dynamic at some point (based on a list of directories/cache layers for instance)